File: abc/240525/d.py
N = int(input())

points = []
for _ in range(N):
    points.append(tuple(map(int, input().split())))

class Node:pass
k = 2

def kdtree(pointList, depth=0):
    if not pointList:
        return

    # 深さに応じて軸を選択し、軸が順次選択されるようにする
    axis = depth % k

    # 点のリストをソートし、中央値の点を選択する
    pointList.sort(key=lambda x:x[axis])
    median = len(pointList)//2 # 中央値を選択

    # ノードを作成し、部分木を構築する
    node = Node()
    node.location = pointList[median]
    node.leftChild = kdtree(pointList[0:median], depth+1)
    node.rightChild = kdtree(pointList[median+1:], depth+1)
    return node

def count_range(node, x_max, y_min, depth=0):
    if node is None:
        return 0
    location = node.location
    axis = depth % k
    value = location[axis]

    if axis == 0:
        # x
        if value > x_max:
            return count_range(node.leftChild, x_max, y_min, depth+1)
        else:
            return count_range(node.leftChild, x_max, y_min, depth+1) + count_range(node.rightChild, x_max, y_min, depth+1) + 1
    else:
        # y
        if value < y_min:
            return count_range(node.rightChild, x_max, y_min, depth+1)
        else:
            return count_range(node.leftChild, x_max, y_min, depth+1) + count_range(node.rightChild, x_max, y_min, depth+1) + 1

rootNode = kdtree(points)

sum = 0
for point in points:
    sum += count_range(rootNode, point[1], point[0])

print((sum - N) // 2)

File: abc/240525/d.cpp
#include <bits/stdc++.h>

using namespace std;


class Node {
    pair<int, int> location;
    shared_ptr<Node> leftChild;
    shared_ptr<Node> rightChild;

public:
    Node(vector<pair<int, int>> &points, int start, int end, int depth) {
        int axis = depth % 2;

        // sort by axis gcc
        sort(points.begin() + start, points.begin() + end, [axis](pair<int, int> a, pair<int, int> b) {
            if (axis == 0) {
                return a.first < b.first;
            } else {
                return a.second < b.second;
            }
        });

        int median = points.size() / 2;
        location = points[median];
        leftChild = make_shared<Node>(points, 0, median, depth + 1);
        rightChild = make_shared<Node>(points, median, points.size(), depth + 1);
    }

    int count_range(int x_max, int y_min, int depth) {
        int axis = depth % 2;

        int value = axis == 0 ? location.first : location.second;
        
        if (axis == 0) {
            if (value > x_max) {
                return leftChild->count_range(x_max, y_min, depth + 1);
            }
            return leftChild->count_range(x_max, y_min, depth + 1) + rightChild->count_range(x_max, y_min, depth + 1) + 1;
        } else {
            if (value < y_min) {
                return rightChild->count_range(x_max, y_min, depth + 1);
            }
            return leftChild->count_range(x_max, y_min, depth + 1) + rightChild->count_range(x_max, y_min, depth + 1) + 1;
        }
    }
};

int main() {
    int N;
    cin >> N;
    vector<pair<int, int>> a(N);

    for (int i = 0; i < N; i++) {
        int x, y;
        cin >> x >> y;
        a.at(i) = make_pair(x, y);
    }

    Node root(a, 0, N, 0);

    int sum = 0;
    for (auto p : a) {
        sum += root.count_range(p.second, p.first, 0);
    }

    cout << (sum - N) / 2;

}